pkg/traversal/traversal.go

```diff
@@ -16,127 +16,63 @@ import (
-	}
-
-	if isManifest {
-		return m.IterateAddresses(ctx, func(manifestNodeAddr swarm.Address) error {
-			return s.processBytes(ctx, manifestNodeAddr, chunkAddressFunc)
-		})
-	}
-	return s.processBytes(ctx, reference, chunkAddressFunc)
+// NewService is a convenient constructor for Service.
+func NewService(store storage.Storer) *Service {
+	return &Service{store: store}
 }
 
-func (s *traversalService) TraverseBytesAddresses(
-	ctx context.Context,
-	reference swarm.Address,
-	chunkAddressFunc swarm.AddressIterFunc,
-) error {
-	return s.processBytes(ctx, reference, chunkAddressFunc)
+// Service is implementation of Interface using storage.Storer as its storage.
+type Service struct {
+	store storage.Storer
 }
 
-func (s *traversalService) TraverseManifestAddresses(
-	ctx context.Context,
-	reference swarm.Address,
-	chunkAddressFunc swarm.AddressIterFunc,
-) error {
-
-	isManifest, m, err := s.checkIsManifest(ctx, reference)
-	if err != nil {
-		return err
-	}
-	if !isManifest {
-		return ErrInvalidType
-	}
-
-	err = m.IterateAddresses(ctx, func(manifestNodeAddr swarm.Address) error {
-		return s.processBytes(ctx, manifestNodeAddr, chunkAddressFunc)
-	})
-	if err != nil {
-		return fmt.Errorf("traversal: iterate chunks: %s: %w", reference, err)
-	}
-
-	return nil
-}
-
-// checkIsManifest checks if the content is manifest.
-func (s *traversalService) checkIsManifest(
-	ctx context.Context,
-	reference swarm.Address,
-) (isManifest bool, m manifest.Interface, err error) {
-
-	// NOTE: 'encrypted' parameter only used for saving manifest
-	m, err = manifest.NewDefaultManifestReference(
-		reference,
-		loadsave.New(s.storer, storage.ModePutRequest, false),
-	)
-	if err != nil {
-		if err == manifest.ErrInvalidManifestType {
-			// ignore
-			err = nil
-			return
+// Traverse implements Traverser.Traverse method.
+func (s *Service) Traverse(ctx context.Context, addr swarm.Address, iterFn swarm.AddressIterFunc) error {
+	processBytes := func(ref swarm.Address) error {
+		j, _, err := joiner.New(ctx, s.store, ref)
+		if err != nil {
+			return fmt.Errorf("traversal: joiner error on %q: %w", ref, err)
+		}
+		err = j.IterateChunkAddresses(iterFn)
+		if err != nil {
+			return fmt.Errorf("traversal: iterate chunk address error for %q: %w", ref, err)
 		}
-		err = fmt.Errorf("traversal: read manifest: %s: %w", reference, err)
-		return
+		return nil
 	}
-	isManifest = true
-	return
-}
 
-func (s *traversalService) processBytes(
-	ctx context.Context,
-	reference swarm.Address,
-	chunkAddressFunc swarm.AddressIterFunc,
-) error {
-	j, _, err := joiner.New(ctx, s.storer, reference)
-	if err != nil {
-		return fmt.Errorf("traversal: joiner: %s: %w", reference, err)
+	ls := loadsave.New(s.store, storage.ModePutRequest, false)
+	switch mf, err := manifest.NewDefaultManifestReference(addr, ls); {
+	case errors.Is(err, manifest.ErrInvalidManifestType):
+		break
+	case err != nil:
+		return fmt.Errorf("traversal: unable to read manifest for %q: %w", addr, err)
```

Comment:
`NewDefaultManifestReference` actually will not read anything until you start using the data structure somehow, so this error will never occur